### PR TITLE
Implement shakeExtra, which is a TypeRep to Dynamic map in ShakeConfig.

### LINKED ISSUE #383: Generalize 'ShakeOptions' argument in Rule's storedValue method
It would be very helpful if 'ShakeOptions' had a user-programmable, 'TypeRep' map which they could use to add their own data to be available when implementing 'storedValue' and 'equalValue'. Otherwise, it's too difficult to implement a Rule instance when the 'storedValue' needs some (non-global) configuration. File gets a config flag; why not our own rules too!

### PR DESCRIPTION
**Pull request by ezyang**
Fixes #383.

Signed-off-by: Edward Z. Yang <ezyang@cs.stanford.edu>

### CODE REVIEW

**Comment by ndmitchell** [comment]
Why not use the existing unordered containers package, which is already a dependency?

File: src/Development/Shake/Types.hs [hunk 1]
```diff
@@ -164,24 +173,27 @@ shakeOptions = ShakeOptions
...
-    ,"shakeLiveFiles","shakeVersionIgnore","shakeProgress", "shakeOutput"]
+    ,"shakeLiveFiles","shakeVersionIgnore","shakeProgress", "shakeOutput", "shakeExtra"]
 tyShakeOptions = mkDataType "Development.Shake.Types.ShakeOptions" [conShakeOptions]
 conShakeOptions = mkConstr tyShakeOptions "ShakeOptions" fieldsShakeOptions Prefix
-unhide x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20 x21 y1 y2 =
-    ShakeOptions x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20 x21 (fromFunction y1) (fromFunction y2)
+unhide x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20 x21 y1 y2 y3 =
+    ShakeOptions x1 x2 x3 x4 x5 x6 x7 x8 x9 x10 x11 x12 x13 x14 x15 x16 x17 x18 x19 x20 x21 (fromFunction y1) (fromFunction y2) (fromFunction y3)
+    -- NB: use of fromFunction for y3 is a little bit of a hack; but
+    -- arguably we should call this fromHide not fromFunction
```
Comment (ndmitchell) [inline src/Development/Shake/Types.hs]: I'm happy renaming Function to Hide - seems entirely reasonable.

**Comment by ndmitchell** [comment]
Thanks for the patch. Looks good!

**Comment by ezyang** [comment]
TypeRep only has an Ord instance, no hashable instance, so we need an Ord-based container.

I wonder why Travis is failing.

**Comment by ndmitchell** [comment]
Probably failing due to https://github.com/ndmitchell/shake/blob/master/src/Test/Self.hs#L87 - that list needs containers adding.

However, I'm not super happy about adding an entire containers dependency just for a a single map. Also, TypeRep does have a Hashable, because I asked for it, or are you thinking of it somehow else?

**Comment by ndmitchell** [comment]
Note the Hashable TypeRep instance is in Data.Hashable, so its not orphan, but if you just have TypeRep in scope at ghci, you won't see it.

**Comment by ezyang** [comment]
Oh I see! Yes, I was looking at the Haddock docs for TypeRep which of course aren't going to report a TypeRep hashable. OK I'll switch it to an unordered container.

**Comment by ezyang** [comment]
I don't know what's going on with the resource busy error in 7.4. Intermittent maybe?

**Comment by ndmitchell** [comment]
Probably, I wouldn't overly worry - GHC 7.4 has some segfaults and race conditions. Shake seems to be hitting this more and more on Travis (perhaps upgraded Travis machines?) so I might add test retrying on older versions.

**Pull request merged**